File: backend/src/services/summary.ts
import { PrismaClient, Category } from '@prisma/client';
import { prisma } from '../config/database';

export interface CategoryBreakdown {
  category: Category;
  amount: number;
  count: number;
}

export interface DailySummary {
  date: string;
  total: number;
  count: number;
  breakdown: CategoryBreakdown[];
}

export interface WeeklySummary {
  week: string;
  startDate: string;
  endDate: string;
  total: number;
  count: number;
  breakdown: CategoryBreakdown[];
  dailyTotals: { date: string; total: number }[];
}

export interface MonthlySummary {
  month: string;
  year: number;
  total: number;
  count: number;
  breakdown: CategoryBreakdown[];
  dailyTotals: { date: string; total: number }[];
}

export const getDailySummary = async (
  userId: string,
  date: Date
): Promise<DailySummary> => {
  const startOfDay = new Date(date);
  startOfDay.setHours(0, 0, 0, 0);
  
  const endOfDay = new Date(date);
  endOfDay.setHours(23, 59, 59, 999);

  const expenses = await prisma.expense.findMany({
    where: {
      userId,
      date: {
        gte: startOfDay,
        lte: endOfDay,
      },
    },
  });

  const total = expenses.reduce((sum, exp) => sum + exp.amount, 0);
  
  const breakdownMap = new Map<Category, { amount: number; count: number }>();
  
  expenses.forEach((exp) => {
    const existing = breakdownMap.get(exp.category) || { amount: 0, count: 0 };
    breakdownMap.set(exp.category, {
      amount: existing.amount + exp.amount,
      count: existing.count + 1,
    });
  });

  const breakdown: CategoryBreakdown[] = Array.from(breakdownMap.entries()).map(
    ([category, data]) => ({
      category,
      amount: data.amount,
      count: data.count,
    })
  );

  return {
    date: date.toISOString().split('T')[0],
    total,
    count: expenses.length,
    breakdown,
  };
};

export const getWeeklySummary = async (
  userId: string,
  startDate: Date,
  endDate: Date
): Promise<WeeklySummary> => {
  const expenses = await prisma.expense.findMany({
    where: {
      userId,
      date: {
        gte: startDate,
        lte: endDate,
      },
    },
    orderBy: {
      date: 'asc',
    },
  });

  const total = expenses.reduce((sum, exp) => sum + exp.amount, 0);

  const breakdownMap = new Map<Category, { amount: number; count: number }>();
  const dailyTotalsMap = new Map<string, number>();

  expenses.forEach((exp) => {
    // Category breakdown
    const existing = breakdownMap.get(exp.category) || { amount: 0, count: 0 };
    breakdownMap.set(exp.category, {
      amount: existing.amount + exp.amount,
      count: existing.count + 1,
    });

    // Daily totals
    const dateKey = exp.date.toISOString().split('T')[0];
    dailyTotalsMap.set(dateKey, (dailyTotalsMap.get(dateKey) || 0) + exp.amount);
  });

  const breakdown: CategoryBreakdown[] = Array.from(breakdownMap.entries()).map(
    ([category, data]) => ({
      category,
      amount: data.amount,
      count: data.count,
    })
  );

  const dailyTotals = Array.from(dailyTotalsMap.entries()).map(([date, total]) => ({
    date,
    total,
  }));

  return {
    week: `${startDate.toISOString().split('T')[0]} to ${endDate.toISOString().split('T')[0]}`,
    startDate: startDate.toISOString().split('T')[0],
    endDate: endDate.toISOString().split('T')[0],
    total,
    count: expenses.length,
    breakdown,
    dailyTotals,
  };
};

export const getMonthlySummary = async (
  userId: string,
  year: number,
  month: number
): Promise<MonthlySummary> => {
  const startDate = new Date(year, month - 1, 1);
  const endDate = new Date(year, month, 0, 23, 59, 59, 999);

  const expenses = await prisma.expense.findMany({
    where: {
      userId,
      date: {
        gte: startDate,
        lte: endDate,
      },
    },
    orderBy: {
      date: 'asc',
    },
  });

  const total = expenses.reduce((sum, exp) => sum + exp.amount, 0);

  const breakdownMap = new Map<Category, { amount: number; count: number }>();
  const dailyTotalsMap = new Map<string, number>();

  expenses.forEach((exp) => {
    // Category breakdown
    const existing = breakdownMap.get(exp.category) || { amount: 0, count: 0 };
    breakdownMap.set(exp.category, {
      amount: existing.amount + exp.amount,
      count: existing.count + 1,
    });

    // Daily totals
    const dateKey = exp.date.toISOString().split('T')[0];
    dailyTotalsMap.set(dateKey, (dailyTotalsMap.get(dateKey) || 0) + exp.amount);
  });

  const breakdown: CategoryBreakdown[] = Array.from(breakdownMap.entries()).map(
    ([category, data]) => ({
      category,
      amount: data.amount,
      count: data.count,
    })
  );

  const dailyTotals = Array.from(dailyTotalsMap.entries()).map(([date, total]) => ({
    date,
    total,
  }));

  const monthNames = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December'
  ];

  return {
    month: monthNames[month - 1],
    year,
    total,
    count: expenses.length,
    breakdown,
    dailyTotals,
  };
};

File: frontend/lib/models/summary.dart
import 'package:pocketa_expense_tracker/models/expense.dart';

class CategoryBreakdown {
  final Category category;
  final double amount;
  final int count;

  CategoryBreakdown({
    required this.category,
    required this.amount,
    required this.count,
  });

  factory CategoryBreakdown.fromJson(Map<String, dynamic> json) {
    return CategoryBreakdown(
      category: Category.fromString(json['category'] as String),
      amount: (json['amount'] as num).toDouble(),
      count: json['count'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category': category.name,
      'amount': amount,
      'count': count,
    };
  }
}

class DailySummary {
  final String date;
  final double total;
  final int count;
  final List<CategoryBreakdown> breakdown;

  DailySummary({
    required this.date,
    required this.total,
    required this.count,
    required this.breakdown,
  });

  factory DailySummary.fromJson(Map<String, dynamic> json) {
    return DailySummary(
      date: json['date'] as String,
      total: (json['total'] as num).toDouble(),
      count: json['count'] as int,
      breakdown: (json['breakdown'] as List)
          .map((e) => CategoryBreakdown.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'date': date,
      'total': total,
      'count': count,
      'breakdown': breakdown.map((e) => e.toJson()).toList(),
    };
  }
}

class WeeklySummary {
  final String week;
  final String startDate;
  final String endDate;
  final double total;
  final int count;
  final List<CategoryBreakdown> breakdown;
  final List<DailyTotal> dailyTotals;

  WeeklySummary({
    required this.week,
    required this.startDate,
    required this.endDate,
    required this.total,
    required this.count,
    required this.breakdown,
    required this.dailyTotals,
  });

  factory WeeklySummary.fromJson(Map<String, dynamic> json) {
    return WeeklySummary(
      week: json['week'] as String,
      startDate: json['startDate'] as String,
      endDate: json['endDate'] as String,
      total: (json['total'] as num).toDouble(),
      count: json['count'] as int,
      breakdown: (json['breakdown'] as List)
          .map((e) => CategoryBreakdown.fromJson(e as Map<String, dynamic>))
          .toList(),
      dailyTotals: (json['dailyTotals'] as List)
          .map((e) => DailyTotal.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class MonthlySummary {
  final String month;
  final int year;
  final double total;
  final int count;
  final List<CategoryBreakdown> breakdown;
  final List<DailyTotal> dailyTotals;

  MonthlySummary({
    required this.month,
    required this.year,
    required this.total,
    required this.count,
    required this.breakdown,
    required this.dailyTotals,
  });

  factory MonthlySummary.fromJson(Map<String, dynamic> json) {
    return MonthlySummary(
      month: json['month'] as String,
      year: json['year'] as int,
      total: (json['total'] as num).toDouble(),
      count: json['count'] as int,
      breakdown: (json['breakdown'] as List)
          .map((e) => CategoryBreakdown.fromJson(e as Map<String, dynamic>))
          .toList(),
      dailyTotals: (json['dailyTotals'] as List)
          .map((e) => DailyTotal.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class DailyTotal {
  final String date;
  final double total;

  DailyTotal({
    required this.date,
    required this.total,
  });

  factory DailyTotal.fromJson(Map<String, dynamic> json) {
    return DailyTotal(
      date: json['date'] as String,
      total: (json['total'] as num).toDouble(),
    );
  }
}
